Fix for changes of Splitter.strip_cls having no effect

CdrSonan opened the pull request:
The current implementation of `kivy.uix.Splitter.strip_cls` only causes `strip_cls` to be instantiated once, when the `_strip` attribute of the Splitter is not yet set. Therefore, any subsequent changes of `strip_cls` have no effect, even though `on_sizable_from()` is called correctly. This also includes setting `strip_cls` in a .kv file, in which case `on_sizable_from()` is called twice: Once with the default strip class, and a second time with the correct one.

As a fix, I am suggesting to remove the else statement preventing `strip_cls` from being instantiated several times. This also matches the code to the implementation of `dropdown_cls` of the Spinner widget, and its refresh mechanism in `kivy.uix.Spinner,_build_dropdown()`.

However, this change also causes re-instantiation of `strip_cls` when the `sizable_from` property is changed. This could theoretically cause loss of data stored in the old instance, but since `sizable_from` is usually only set once during UI creation, this is unlikely to be a problem in practice.
A possible solution for this issue would be to check whether `strip_cls` matches the class of the current `_strip` attribute, and only perform the re-instantiation if that is not the case.
If this is worth the additional complexity, I can update this pull request accordingly.

Maintainer merge checklist
* [ ] Title is descriptive/clear for inclusion in release notes.
* [x] Applied a `Component: xxx` label.
* [ ] Applied the `api-deprecation` or `api-break` label.
* [ ] Applied the `release-highlight` label to be highlighted in release notes.
* [ ] Added to the milestone version it was merged into.
* [ ] **Unittests** are included in PR.
* [ ] Properly documented, including `versionadded`, `versionchanged` as needed.

misl6 commented:
> However, this change also causes re-instantiation of strip_cls when the sizable_from property is changed. This could theoretically cause loss of data stored in the old instance, but since sizable_from is usually only set once during UI creation, this is unlikely to be a problem in practice.
A possible solution for this issue would be to check whether strip_cls matches the class of the current _strip attribute, and only perform the re-instantiation if that is not the case.
If this is worth the additional complexity, I can update this pull request accordingly.

Maybe we can:
- Separate the binding / unbinding / size change logic into a separate method
- Update `instance._strip` when `instance.strip_cls` changes + trigger the binding / unbinding logic after that
- Trigger the binding / unbinding logic when `sizable_from` changes

Also, FYI, here we have the opportunity to remove `from kivy.compat import string_types` 😀

Let me know if you want to enhance this PR or if you're fine with it.

CdrSonan commented:
Thanks for your reply!
I am going to implement what you said and then update the pull request.

misl6 commented:
> Thanks for your reply!
> I am going to implement what you said and then update the pull request.

Love it. Feel free to ping me by requesting a review when you're ready!

CdrSonan commented:
@misl6 
I've finally had time to get back to this.
Moving the reinstancing logic to a new function hasn't proven practical. Most parts of `on_sizable_from()` would also be necessary in the hypothetical `on_strip_cls()`, so there would either be duplicate code between the two, or additional utility functions for shared functionality. (Mainly various callback binds/unbinds)

So I have instead replaced the original else statement with an `isinstance()` call that checks whether `instance._strip` is indeed an instance of `strip_cls`. If it isn't, the strip is reinstanced, otherwise the reinstancing code is skipped. Since `None` is not an instance of `strip_cls`, this also means reinstancing is triggered if the strip hasn't been instanced yet,
Therefore, everything works as it's supposed to. Reinstancing is triggered when the widget is created or when `strip_cls` changes, but skipped when only `sizable_from` changes.
I have tested this implementation in both the example Splitter App and my own code using custom strip classes, and encountered no issues.

I have also removed the `string_types` import and replaced it with just `str`. I am not sure if this has the desired effect since it doesn't offer any additional backwards compatibility, but it is equivalent to the original code, given the current implementation of `string_types`.

misl6 reviewed (approved):
LGTM. Thank you!

misl6 commented:
> I have also removed the string_types import and replaced it with just str. I am not sure if this has the desired effect since it doesn't offer any additional backwards compatibility, but it is equivalent to the original code, given the current implementation of string_types.

We still have a few `string_types` around from the Python 2 era, but we removed Python 2 compatibility a long time ago. We take the opportunity to clean up the code when possible 😀
